// @solana/kit-plugin-signer/dist/index.node.cjs
'use strict';

var kit = require('@solana/kit');
var fs = require('fs');

// src/airdrop-signer.ts
function airdropSigner(amount) {
  return async (client) => {
    if (client.payer !== client.identity) {
      throw new Error(
        "`airdropSigner` requires the payer and identity to be the same signer on the client. Use `airdropPayer` and/or `airdropIdentity` if you want to airdrop to the payer and/or identity, respectively."
      );
    }
    if (!client.airdrop) {
      throw new Error("`airdropSigner` requires an airdrop function on the client");
    }
    await client.airdrop(client.payer.address, amount);
    return client;
  };
}
function airdropPayer(amount) {
  return async (client) => {
    if (!client.airdrop) {
      throw new Error("`airdropPayer` requires an airdrop function on the client");
    }
    await client.airdrop(client.payer.address, amount);
    return client;
  };
}
function airdropIdentity(amount) {
  return async (client) => {
    if (!client.airdrop) {
      throw new Error("`airdropIdentity` requires an airdrop function on the client");
    }
    await client.airdrop(client.identity.address, amount);
    return client;
  };
}
function signer(signer2) {
  return (client) => kit.extendClient(client, { identity: signer2, payer: signer2 });
}
function payer(payer2) {
  return (client) => kit.extendClient(client, { payer: payer2 });
}
function identity(identity2) {
  return (client) => kit.extendClient(client, { identity: identity2 });
}

// src/generated-signer.ts
function generatedSigner() {
  return async (client) => signer(await kit.generateKeyPairSigner())(client);
}
function generatedPayer() {
  return async (client) => payer(await kit.generateKeyPairSigner())(client);
}
function generatedIdentity() {
  return async (client) => identity(await kit.generateKeyPairSigner())(client);
}
function generatedSignerWithSol(amount) {
  return async (client) => await kit.pipe(client, generatedSigner(), async (client2) => await airdropSigner(amount)(await client2));
}
function generatedPayerWithSol(amount) {
  return async (client) => await kit.pipe(client, generatedPayer(), async (client2) => await airdropPayer(amount)(await client2));
}
function generatedIdentityWithSol(amount) {
  return async (client) => await kit.pipe(client, generatedIdentity(), async (client2) => await airdropIdentity(amount)(await client2));
}
function signerFromFile(path) {
  return async (client) => signer(await getKeyPairSignerFromFile(path))(client);
}
function payerFromFile(path) {
  return async (client) => payer(await getKeyPairSignerFromFile(path))(client);
}
function identityFromFile(path) {
  return async (client) => identity(await getKeyPairSignerFromFile(path))(client);
}
async function getKeyPairSignerFromFile(path) {
  const bytes = JSON.parse(fs.readFileSync(path, "utf-8"));
  return await kit.createKeyPairSignerFromBytes(new Uint8Array(bytes));
}

exports.airdropIdentity = airdropIdentity;
exports.airdropPayer = airdropPayer;
exports.airdropSigner = airdropSigner;
exports.generatedIdentity = generatedIdentity;
exports.generatedIdentityWithSol = generatedIdentityWithSol;
exports.generatedPayer = generatedPayer;
exports.generatedPayerWithSol = generatedPayerWithSol;
exports.generatedSigner = generatedSigner;
exports.generatedSignerWithSol = generatedSignerWithSol;
exports.identity = identity;
exports.identityFromFile = identityFromFile;
exports.payer = payer;
exports.payerFromFile = payerFromFile;
exports.signer = signer;
exports.signerFromFile = signerFromFile;
