// @solana/kit-plugin-signer/dist/index.browser.mjs
import { extendClient, generateKeyPairSigner, pipe } from '@solana/kit';

// src/airdrop-signer.ts
function airdropSigner(amount) {
  return async (client) => {
    if (client.payer !== client.identity) {
      throw new Error(
        "`airdropSigner` requires the payer and identity to be the same signer on the client. Use `airdropPayer` and/or `airdropIdentity` if you want to airdrop to the payer and/or identity, respectively."
      );
    }
    if (!client.airdrop) {
      throw new Error("`airdropSigner` requires an airdrop function on the client");
    }
    await client.airdrop(client.payer.address, amount);
    return client;
  };
}
function airdropPayer(amount) {
  return async (client) => {
    if (!client.airdrop) {
      throw new Error("`airdropPayer` requires an airdrop function on the client");
    }
    await client.airdrop(client.payer.address, amount);
    return client;
  };
}
function airdropIdentity(amount) {
  return async (client) => {
    if (!client.airdrop) {
      throw new Error("`airdropIdentity` requires an airdrop function on the client");
    }
    await client.airdrop(client.identity.address, amount);
    return client;
  };
}
function signer(signer2) {
  return (client) => extendClient(client, { identity: signer2, payer: signer2 });
}
function payer(payer2) {
  return (client) => extendClient(client, { payer: payer2 });
}
function identity(identity2) {
  return (client) => extendClient(client, { identity: identity2 });
}

// src/generated-signer.ts
function generatedSigner() {
  return async (client) => signer(await generateKeyPairSigner())(client);
}
function generatedPayer() {
  return async (client) => payer(await generateKeyPairSigner())(client);
}
function generatedIdentity() {
  return async (client) => identity(await generateKeyPairSigner())(client);
}
function generatedSignerWithSol(amount) {
  return async (client) => await pipe(client, generatedSigner(), async (client2) => await airdropSigner(amount)(await client2));
}
function generatedPayerWithSol(amount) {
  return async (client) => await pipe(client, generatedPayer(), async (client2) => await airdropPayer(amount)(await client2));
}
function generatedIdentityWithSol(amount) {
  return async (client) => await pipe(client, generatedIdentity(), async (client2) => await airdropIdentity(amount)(await client2));
}
function signerFromFile(path) {
  {
    throw new Error("`signerFromFile` is only supported in Node.js environments");
  }
}
function payerFromFile(path) {
  {
    throw new Error("`payerFromFile` is only supported in Node.js environments");
  }
}
function identityFromFile(path) {
  {
    throw new Error("`identityFromFile` is only supported in Node.js environments");
  }
}

export { airdropIdentity, airdropPayer, airdropSigner, generatedIdentity, generatedIdentityWithSol, generatedPayer, generatedPayerWithSol, generatedSigner, generatedSignerWithSol, identity, identityFromFile, payer, payerFromFile, signer, signerFromFile };
