// @solana/kit-plugin-instruction-plan/src/core.ts
import {
    assertIsSingleTransactionPlan,
    assertIsSingleTransactionPlanResult,
    assertIsSuccessfulSingleTransactionPlanResult,
    CanceledSingleTransactionPlanResult,
    ClientWithTransactionPlanning,
    ClientWithTransactionSending,
    createFailedToSendTransactionError,
    createFailedToSendTransactionsError,
    extendClient,
    FailedSingleTransactionPlanResult,
    isSolanaError,
    isTransactionPlan,
    parseInstructionOrTransactionPlanInput,
    parseInstructionPlanInput,
    singleTransactionPlan,
    SOLANA_ERROR__INSTRUCTION_PLANS__FAILED_TO_EXECUTE_TRANSACTION_PLAN,
    TransactionPlanExecutor,
    TransactionPlanner,
    TransactionPlanResult,
} from '@solana/kit';

/**
 * A plugin that sets the transactionPlanner on the client.
 *
 * @example
 * ```ts
 * import { createClient, createTransactionPlanner } from '@solana/kit';
 * import { transactionPlanner } from '@solana/kit-plugin-instruction-plan';
 *
 * // Install the transactionPlanner plugin using a custom transaction planner.
 * const client = createClient()
 *     .use(transactionPlanner(createTransactionPlanner(...)));
 *
 * // Use the transaction planner.
 * const transactionPlan = await client.transactionPlanner(myInstructionPlan);
 * ```
 */
export function transactionPlanner(transactionPlanner: TransactionPlanner) {
    return <T extends object>(client: T) => extendClient(client, { transactionPlanner });
}

/**
 * A plugin that sets the transactionPlanExecutor on the client.
 *
 * @example
 * ```ts
 * import { createClient, createTransactionPlanExecutor } from '@solana/kit';
 * import { transactionPlanExecutor } from '@solana/kit-plugin-instruction-plan';
 *
 * // Install the transactionPlanExecutor plugin using a custom transaction plan executor.
 * const client = createClient()
 *     .use(transactionPlanExecutor(createTransactionPlanExecutor(...)));
 *
 * // Use the transaction plan executor.
 * const transactionPlanResult = await client.transactionPlanExecutor(myTransactionPlan);
 * ```
 */
export function transactionPlanExecutor(transactionPlanExecutor: TransactionPlanExecutor) {
    return <T extends object>(client: T) => extendClient(client, { transactionPlanExecutor });
}

/**
 * A plugin that adds `planTransaction`, `planTransactions`, `sendTransaction` and
 * `sendTransactions` functions on the client to plan and execute transaction messages,
 * instructions or instruction plans.
 *
 * This expects the client to have both a `transactionPlanner`
 * and a `transactionPlanExecutor` set.
 *
 * The `planTransaction` and `planTransactions` functions plan instructions into
 * transaction plans without executing them. The `sendTransaction` and `sendTransactions`
 * functions combine planning and execution in a single call.
 *
 * Note that the `sendTransaction` function will assert that the transaction plan result
 * is both successful and contains a single transaction plan. This is slightly different from
 * the `sendTransactions` function which will return the full transaction plan result
 * as produced by the `transactionPlanExecutor`.
 *
 * @example
 * ```ts
 * import { createClient } from '@solana/kit';
 * import { transactionPlanner, transactionPlanExecutor, planAndSendTransactions } from '@solana/kit-plugin-instruction-plan';
 *
 * // Install the planAndSendTransactions plugin and its requirements.
 * const client = createClient()
 *     .use(transactionPlanner(myTransactionPlanner))
 *     .use(transactionPlanExecutor(myTransactionPlanExecutor))
 *     .use(planAndSendTransactions());
 *
 * // Plan transactions without executing them.
 * const transactionPlan = await client.planTransactions(myInstructionPlan);
 * const transactionMessage = await client.planTransaction(myInstructionPlan);
 *
 * // Plan and execute transactions in one call.
 * const singleResult = await client.sendTransaction(myInstructionPlan);
 * const result = await client.sendTransactions(myInstructionPlan);
 * ```
 */
export function planAndSendTransactions() {
    return <T extends { transactionPlanExecutor: TransactionPlanExecutor; transactionPlanner: TransactionPlanner }>(
        client: T,
    ) => extendClient(client, getTransactionPlanningAndSendingFunctions(client));
}

function getTransactionPlanningAndSendingFunctions(client: {
    transactionPlanExecutor: TransactionPlanExecutor;
    transactionPlanner: TransactionPlanner;
}): ClientWithTransactionPlanning & ClientWithTransactionSending {
    const planTransactions: ClientWithTransactionPlanning['planTransactions'] = async (input, config = {}) => {
        const instructionPlan = parseInstructionPlanInput(input);
        config?.abortSignal?.throwIfAborted();
        return await client.transactionPlanner(instructionPlan, config);
    };

    const planTransaction: ClientWithTransactionPlanning['planTransaction'] = async (input, config = {}) => {
        const transactionPlan = await planTransactions(input, config);
        assertIsSingleTransactionPlan(transactionPlan);
        return transactionPlan.message;
    };

    const sendTransactions: ClientWithTransactionSending['sendTransactions'] = async (input, config = {}) => {
        const plan = parseInstructionOrTransactionPlanInput(input);
        config?.abortSignal?.throwIfAborted();
        const transactionPlan = isTransactionPlan(plan) ? plan : await planTransactions(plan, config);
        config?.abortSignal?.throwIfAborted();
        try {
            return await client.transactionPlanExecutor(transactionPlan, config);
        } catch (error) {
            if (!isSolanaError(error, SOLANA_ERROR__INSTRUCTION_PLANS__FAILED_TO_EXECUTE_TRANSACTION_PLAN)) {
                throw error;
            }
            throw createFailedToSendTransactionsError(
                error.context.transactionPlanResult as TransactionPlanResult,
                error.context.abortReason,
            );
        }
    };

    const sendTransaction: ClientWithTransactionSending['sendTransaction'] = async (input, config = {}) => {
        const plan = parseInstructionOrTransactionPlanInput(input);
        config?.abortSignal?.throwIfAborted();
        const transactionPlan = isTransactionPlan(plan)
            ? plan
            : singleTransactionPlan(await planTransaction(plan, config));
        config?.abortSignal?.throwIfAborted();
        try {
            const result = await client.transactionPlanExecutor(transactionPlan, config);
            assertIsSuccessfulSingleTransactionPlanResult(result);
            return result;
        } catch (error) {
            if (!isSolanaError(error, SOLANA_ERROR__INSTRUCTION_PLANS__FAILED_TO_EXECUTE_TRANSACTION_PLAN)) {
                throw error;
            }
            assertIsSingleTransactionPlanResult(error.context.transactionPlanResult as TransactionPlanResult);
            throw createFailedToSendTransactionError(
                error.context.transactionPlanResult as
                    | CanceledSingleTransactionPlanResult
                    | FailedSingleTransactionPlanResult,
                error.context.abortReason,
            );
        }
    };

    return { planTransaction, planTransactions, sendTransaction, sendTransactions };
}
