// @solana/kit-plugin-rpc/dist/index.browser.mjs
import { airdropFactory, extendClient, BASE_ACCOUNT_SIZE, lamports, sendAndConfirmTransactionFactory, estimateComputeUnitLimitFactory, createTransactionPlanExecutor, pipe, setTransactionMessageLifetimeUsingBlockhash, signTransactionMessageWithSigners, assertIsTransactionWithBlockhashLifetime, getTransactionMessageComputeUnitLimit, isSolanaError, SOLANA_ERROR__TRANSACTION__FAILED_WHEN_SIMULATING_TO_ESTIMATE_COMPUTE_LIMIT, setTransactionMessageComputeUnitLimit, createTransactionPlanner, createTransactionMessage, setTransactionMessageFeePayerSigner, fillTransactionMessageProvisoryComputeUnitLimit, setTransactionMessageComputeUnitPrice, createSolanaRpc, createSolanaRpcSubscriptions } from '@solana/kit';
import { planAndSendTransactions } from '@solana/kit-plugin-instruction-plan';

// src/airdrop.ts
function rpcAirdrop() {
  return (client) => {
    const airdropInternal = airdropFactory({ rpc: client.rpc, rpcSubscriptions: client.rpcSubscriptions });
    return extendClient(client, {
      airdrop: (address, amount, abortSignal) => airdropInternal({ abortSignal, commitment: "confirmed", lamports: amount, recipientAddress: address })
    });
  };
}
function rpcGetMinimumBalance() {
  return (client) => {
    return extendClient(client, {
      getMinimumBalance: async (space, config) => {
        if (config?.withoutHeader) {
          const headerBalance = await client.rpc.getMinimumBalanceForRentExemption(0n).send();
          const lamportsPerByte = BigInt(headerBalance) / BigInt(BASE_ACCOUNT_SIZE);
          return lamports(lamportsPerByte * BigInt(space));
        }
        return await client.rpc.getMinimumBalanceForRentExemption(BigInt(space)).send();
      }
    });
  };
}
function rpcConnection(rpc2) {
  return (client) => extendClient(client, { rpc: rpc2 });
}
function rpcSubscriptionsConnection(rpcSubscriptions) {
  return (client) => extendClient(client, { rpcSubscriptions });
}
var MAX_COMPUTE_UNIT_LIMIT = 14e5;
function rpcTransactionPlanExecutor(config = {}) {
  return (client) => {
    if (!client.rpc || !client.rpcSubscriptions) {
      throw new Error(
        "An RPC instance with subscriptions is required on the client to create the RPC transaction plan executor. Please add the RPC plugin to your client before using this plugin."
      );
    }
    const sendAndConfirmTransaction = sendAndConfirmTransactionFactory({
      rpc: client.rpc,
      rpcSubscriptions: client.rpcSubscriptions
    });
    const estimateCULimit = estimateComputeUnitLimitFactory({ rpc: client.rpc });
    const transactionPlanExecutor = createTransactionPlanExecutor({
      executeTransactionMessage: limitFunction(async (context, transactionMessage, executorConfig) => {
        const needsCuEstimation = needsComputeUnitEstimation(transactionMessage);
        const { value: latestBlockhash } = await client.rpc.getLatestBlockhash().send(executorConfig);
        const signedTransaction = await pipe(
          transactionMessage,
          (tx) => setTransactionMessageLifetimeUsingBlockhash(latestBlockhash, tx),
          (tx) => context.message = tx,
          async (tx) => needsCuEstimation ? await estimateAndSetComputeUnitLimit(
            tx,
            estimateCULimit,
            config.skipPreflight ?? false,
            executorConfig
          ) : tx,
          async (tx) => context.message = await tx,
          async (tx) => await signTransactionMessageWithSigners(await tx, executorConfig),
          async (tx) => context.transaction = await tx
        );
        assertIsTransactionWithBlockhashLifetime(signedTransaction);
        await sendAndConfirmTransaction(signedTransaction, {
          commitment: "confirmed",
          skipPreflight: config.skipPreflight || needsCuEstimation,
          ...executorConfig
        });
        return signedTransaction;
      }, config.maxConcurrency ?? 10)
    });
    return extendClient(client, { transactionPlanExecutor });
  };
}
function needsComputeUnitEstimation(transactionMessage) {
  const computeUnitLimit = getTransactionMessageComputeUnitLimit(transactionMessage);
  return !computeUnitLimit || computeUnitLimit === MAX_COMPUTE_UNIT_LIMIT;
}
async function estimateAndSetComputeUnitLimit(transactionMessage, estimateCULimit, skipPreflight, config) {
  let estimatedUnits;
  try {
    estimatedUnits = await estimateCULimit(transactionMessage, config);
  } catch (error) {
    if (skipPreflight && isSolanaError(error, SOLANA_ERROR__TRANSACTION__FAILED_WHEN_SIMULATING_TO_ESTIMATE_COMPUTE_LIMIT)) {
      const bigintUnits = error.context.unitsConsumed ?? 0n;
      estimatedUnits = bigintUnits > 4294967295n ? 4294967295 : Number(bigintUnits);
    } else {
      throw error;
    }
  }
  const units = Math.ceil(estimatedUnits * 1.1);
  return setTransactionMessageComputeUnitLimit(units, transactionMessage);
}
function limitFunction(fn, maxConcurrency) {
  let running = 0;
  const queue = [];
  function process() {
    if (running >= maxConcurrency || queue.length === 0) return;
    running++;
    const { args, resolve, reject } = queue.shift();
    Promise.resolve(fn(...args)).then(resolve).catch(reject).finally(() => {
      running--;
      process();
    });
  }
  return function(...args) {
    return new Promise((resolve, reject) => {
      queue.push({ args, reject, resolve });
      process();
    });
  };
}
function rpcTransactionPlanner(config = {}) {
  return (client) => {
    const transactionPlanner = createTransactionPlanner({
      createTransactionMessage: () => {
        return pipe(
          createTransactionMessage({ version: config.version ?? 0 }),
          (tx) => setTransactionMessageFeePayerSigner(client.payer, tx),
          (tx) => fillTransactionMessageProvisoryComputeUnitLimit(tx),
          (tx) => config.microLamportsPerComputeUnit ? setTransactionMessageComputeUnitPrice(config.microLamportsPerComputeUnit, tx) : tx
        );
      }
    });
    return extendClient(client, { transactionPlanner });
  };
}

// src/solana-rpc.ts
function solanaRpc(config) {
  return (client) => pipe(
    client,
    solanaRpcConnection(config),
    rpcGetMinimumBalance(),
    rpcTransactionPlanner(config.transactionConfig),
    rpcTransactionPlanExecutor({ maxConcurrency: config.maxConcurrency, skipPreflight: config.skipPreflight }),
    planAndSendTransactions()
  );
}
function solanaMainnetRpc(config) {
  return (client) => pipe(client, solanaRpc(config));
}
function solanaDevnetRpc(config) {
  return (client) => pipe(
    client,
    solanaRpc({
      ...config,
      rpcUrl: config?.rpcUrl ?? "https://api.devnet.solana.com"
    }),
    rpcAirdrop()
  );
}
var LOCAL_VALIDATOR_RPC_URL = "http://127.0.0.1:8899";
var LOCAL_VALIDATOR_RPC_SUBSCRIPTIONS_URL = "ws://127.0.0.1:8900";
var LOCAL_VALIDATOR_RPC_SUBSCRIPTIONS_URLS = {
  [LOCAL_VALIDATOR_RPC_URL]: LOCAL_VALIDATOR_RPC_SUBSCRIPTIONS_URL,
  "http://localhost:8899": "ws://localhost:8900"
};
function deriveRpcSubscriptionsUrl(rpcUrl) {
  return LOCAL_VALIDATOR_RPC_SUBSCRIPTIONS_URLS[rpcUrl] ?? rpcUrl.replace(/^http/, "ws");
}
function solanaLocalRpc(config) {
  return (client) => pipe(
    client,
    solanaRpc({
      ...config,
      rpcSubscriptionsUrl: config?.rpcSubscriptionsUrl ?? LOCAL_VALIDATOR_RPC_SUBSCRIPTIONS_URL,
      rpcUrl: config?.rpcUrl ?? LOCAL_VALIDATOR_RPC_URL
    }),
    rpcAirdrop()
  );
}
function solanaRpcConnection(config) {
  const rpc2 = createSolanaRpc(config.rpcUrl, config.rpcConfig);
  const rpcSubscriptionsUrl = config.rpcSubscriptionsUrl ?? deriveRpcSubscriptionsUrl(config.rpcUrl);
  const rpcSubscriptions = createSolanaRpcSubscriptions(
    rpcSubscriptionsUrl,
    config.rpcSubscriptionsConfig
  );
  return (client) => extendClient(client, { rpc: rpc2, rpcSubscriptions });
}
function solanaRpcSubscriptionsConnection(url, config) {
  return rpcSubscriptionsConnection(createSolanaRpcSubscriptions(url, config));
}
function rpc(url, rpcSubscriptionsConfig) {
  const rpc2 = createSolanaRpc(url);
  const rpcSubscriptionsUrl = rpcSubscriptionsConfig?.url ?? deriveRpcSubscriptionsUrl(url);
  const rpcSubscriptions = createSolanaRpcSubscriptions(rpcSubscriptionsUrl, rpcSubscriptionsConfig);
  return (client) => extendClient(client, { rpc: rpc2, rpcSubscriptions });
}
function localhostRpc(url, rpcSubscriptionsConfig) {
  return rpc(
    url ?? LOCAL_VALIDATOR_RPC_URL,
    rpcSubscriptionsConfig ?? { url: LOCAL_VALIDATOR_RPC_SUBSCRIPTIONS_URL }
  );
}

export { localhostRpc, rpc, rpcAirdrop, rpcConnection, rpcGetMinimumBalance, rpcSubscriptionsConnection, rpcTransactionPlanExecutor, rpcTransactionPlanner, solanaDevnetRpc, solanaLocalRpc, solanaMainnetRpc, solanaRpc, solanaRpcConnection, solanaRpcSubscriptionsConnection };
