// @solana/kit-plugin-rpc/src/transaction-plan-executor.ts
import {
    assertIsTransactionWithBlockhashLifetime,
    ClientWithRpc,
    ClientWithRpcSubscriptions,
    createTransactionPlanExecutor,
    estimateComputeUnitLimitFactory,
    extendClient,
    GetEpochInfoApi,
    GetLatestBlockhashApi,
    GetSignatureStatusesApi,
    getTransactionMessageComputeUnitLimit,
    isSolanaError,
    pipe,
    sendAndConfirmTransactionFactory,
    SendTransactionApi,
    setTransactionMessageComputeUnitLimit,
    setTransactionMessageLifetimeUsingBlockhash,
    SignatureNotificationsApi,
    signTransactionMessageWithSigners,
    SimulateTransactionApi,
    SlotNotificationsApi,
    SOLANA_ERROR__TRANSACTION__FAILED_WHEN_SIMULATING_TO_ESTIMATE_COMPUTE_LIMIT,
    TransactionPlanExecutorConfig,
} from '@solana/kit';

const MAX_COMPUTE_UNIT_LIMIT = 1_400_000;

/**
 * A plugin that provides a default transaction plan executor using RPC.
 *
 * The executor handles compute unit estimation, transaction signing, and
 * sending via RPC. A concurrency limit can be set to avoid hitting rate
 * limits when sending many transactions in parallel.
 *
 * @param config - Optional configuration for the executor.
 * @returns A plugin that adds `transactionPlanExecutor` to the client.
 *
 * @example
 * ```ts
 * import { createClient } from '@solana/kit';
 * import { solanaRpcConnection, rpcTransactionPlanner, rpcTransactionPlanExecutor } from '@solana/kit-plugin-rpc';
 * import { generatedPayer } from '@solana/kit-plugin-signer';
 *
 * const client = await createClient()
 *     .use(solanaRpcConnection({ rpcUrl: 'https://api.mainnet-beta.solana.com' }))
 *     .use(generatedPayer())
 *     .use(rpcTransactionPlanner())
 *     .use(rpcTransactionPlanExecutor());
 * ```
 */
export function rpcTransactionPlanExecutor(
    config: {
        /**
         * The maximum number of concurrent executions allowed.
         * Defaults to 10.
         */
        maxConcurrency?: number;
        /**
         * Whether to skip the preflight simulation when sending transactions.
         *
         * When `false` (default), preflight is skipped only if a compute unit
         * estimation simulation was already performed for that transaction.
         * If the transaction has an explicit compute unit limit (i.e. no
         * estimation was needed), preflight runs as the only simulation.
         *
         * When `true`, preflight is always skipped and the transaction is sent
         * directly to the validator. Additionally, if the compute unit estimation
         * simulation fails, the consumed units from the failed simulation are used
         * to set the compute unit limit so the transaction still reaches the
         * validator. This is useful for debugging failed transactions in an explorer.
         *
         * Defaults to `false`.
         */
        skipPreflight?: boolean;
    } = {},
) {
    return <
        T extends ClientWithRpc<
            GetEpochInfoApi &
                GetLatestBlockhashApi &
                GetSignatureStatusesApi &
                SendTransactionApi &
                SimulateTransactionApi
        > &
            ClientWithRpcSubscriptions<SignatureNotificationsApi & SlotNotificationsApi>,
    >(
        client: T,
    ) => {
        if (!client.rpc || !client.rpcSubscriptions) {
            throw new Error(
                'An RPC instance with subscriptions is required on the client to create the RPC transaction plan executor. ' +
                    'Please add the RPC plugin to your client before using this plugin.',
            );
        }

        const sendAndConfirmTransaction = sendAndConfirmTransactionFactory({
            rpc: client.rpc,
            rpcSubscriptions: client.rpcSubscriptions,
        });
        const estimateCULimit = estimateComputeUnitLimitFactory({ rpc: client.rpc });

        const transactionPlanExecutor = createTransactionPlanExecutor({
            executeTransactionMessage: limitFunction(async (context, transactionMessage, executorConfig) => {
                const needsCuEstimation = needsComputeUnitEstimation(transactionMessage);
                const { value: latestBlockhash } = await client.rpc.getLatestBlockhash().send(executorConfig);
                const signedTransaction = await pipe(
                    transactionMessage,
                    tx => setTransactionMessageLifetimeUsingBlockhash(latestBlockhash, tx),
                    tx => (context.message = tx),
                    async tx =>
                        needsCuEstimation
                            ? await estimateAndSetComputeUnitLimit(
                                  tx,
                                  estimateCULimit,
                                  config.skipPreflight ?? false,
                                  executorConfig,
                              )
                            : tx,
                    async tx => (context.message = await tx),
                    async tx => await signTransactionMessageWithSigners(await tx, executorConfig),
                    async tx => (context.transaction = await tx),
                );
                assertIsTransactionWithBlockhashLifetime(signedTransaction);
                await sendAndConfirmTransaction(signedTransaction, {
                    commitment: 'confirmed',
                    skipPreflight: config.skipPreflight || needsCuEstimation,
                    ...executorConfig,
                });
                return signedTransaction;
            }, config.maxConcurrency ?? 10),
        } as TransactionPlanExecutorConfig);

        return extendClient(client, { transactionPlanExecutor });
    };
}

/**
 * Checks whether a transaction message needs compute unit estimation.
 *
 * Estimation is needed when the transaction has no `SetComputeUnitLimit`
 * instruction, or when it has a provisory (`0`) or maximum (`1,400,000`)
 * compute unit limit.
 *
 * @param transactionMessage - The transaction message to check.
 * @returns `true` if the transaction needs compute unit estimation, `false` otherwise.
 */
function needsComputeUnitEstimation(
    transactionMessage: Parameters<typeof getTransactionMessageComputeUnitLimit>[0],
): boolean {
    const computeUnitLimit = getTransactionMessageComputeUnitLimit(transactionMessage);
    return !computeUnitLimit || computeUnitLimit === MAX_COMPUTE_UNIT_LIMIT;
}

/**
 * Estimates the compute unit limit for a transaction message via simulation
 * and sets the result on the message with a 10% buffer.
 *
 * When `skipPreflight` is `true` and the estimation simulation fails, the consumed
 * units from the failed simulation are used so the transaction can still reach the
 * validator for debugging purposes.
 *
 * @param transactionMessage - The transaction message to estimate and set the compute unit limit on.
 * @param estimateCULimit - A function that estimates the compute unit limit via simulation.
 * @param skipPreflight - Whether to recover from failed simulations using consumed units.
 * @param config - Optional configuration forwarded to the estimator (e.g. abort signal).
 * @returns The updated transaction message with the estimated compute unit limit.
 */
async function estimateAndSetComputeUnitLimit<
    TMessage extends Parameters<typeof setTransactionMessageComputeUnitLimit>[1],
>(
    transactionMessage: TMessage,
    estimateCULimit: (tx: TMessage, config?: { abortSignal?: AbortSignal }) => Promise<number>,
    skipPreflight: boolean,
    config?: { abortSignal?: AbortSignal },
) {
    let estimatedUnits;
    try {
        estimatedUnits = await estimateCULimit(transactionMessage, config);
    } catch (error) {
        if (
            skipPreflight &&
            isSolanaError(error, SOLANA_ERROR__TRANSACTION__FAILED_WHEN_SIMULATING_TO_ESTIMATE_COMPUTE_LIMIT)
        ) {
            // Use consumed units from the failed simulation so the
            // transaction can still reach the validator for debugging.
            // The unitsConsumed field is a raw bigint from the RPC response,
            // so we downcast it to a u32 number, capping at 4_294_967_295.
            const bigintUnits = error.context.unitsConsumed ?? 0n;
            estimatedUnits = bigintUnits > 4_294_967_295n ? 4_294_967_295 : Number(bigintUnits);
        } else {
            throw error;
        }
    }

    // Multiply the simulated limit by 1.1 to add a 10% buffer.
    const units = Math.ceil(estimatedUnits * 1.1);
    return setTransactionMessageComputeUnitLimit(units, transactionMessage);
}

/**
 * Limits the number of concurrent executions of an asynchronous function.
 *
 * This utility creates a wrapper around an async function that enforces
 * a maximum concurrency limit. When the limit is reached, additional
 * calls are queued and executed when capacity becomes available.
 *
 * @param fn - The asynchronous function to be limited.
 * @param maxConcurrency - The maximum number of concurrent executions allowed.
 * @returns A new function that enforces the concurrency limit.
 *
 * @example
 * ```ts
 * const limitedFetchData = limitFunction(fetchData, 2);
 *
 * // Only 2 fetchData calls will run concurrently.
 * const results = await Promise.all([
 *   limitedFetchData('url1'),
 *   limitedFetchData('url2'),
 *   limitedFetchData('url1'), // This will wait until one of the previous calls completes.
 * ]);
 * ```
 */
function limitFunction<TArguments extends unknown[], TReturnType>(
    fn: (...args: TArguments) => PromiseLike<TReturnType>,
    maxConcurrency: number,
): (...args: TArguments) => Promise<TReturnType> {
    let running = 0;
    const queue: Array<{
        args: TArguments;
        reject: (reason?: unknown) => void;
        resolve: (value: TReturnType) => void;
    }> = [];

    function process() {
        // Do nothing if we're still running at max concurrency
        // or if there's nothing left to process.
        if (running >= maxConcurrency || queue.length === 0) return;

        running++;
        const { args, resolve, reject } = queue.shift()!;

        Promise.resolve(fn(...args))
            .then(resolve)
            .catch(reject)
            .finally(() => {
                running--;
                process();
            });
    }

    return function (...args) {
        return new Promise((resolve, reject) => {
            queue.push({ args, reject, resolve });
            process();
        });
    };
}
